src/run.rs

```diff
@@ -56,14 +68,14 @@ fn iterate_files(files: Vec<String>) -> impl Iterator<Item = PathBuf> {
                     // filter out directories
                     && !d.file_type().is_dir()
             }
-            Err(e) => {
-                println!("Unable to read file: {}", e);
-                false
-            }
+            Err(_) => true,
         })
-        .map(|f| {
-            f.expect("This should not happen: we have already filtered out the errors")
-                .into_path()
+        .map(|r| match r {
+            Ok(d) => Ok(d.into_path()),
+            Err(e) => match e.path() {
+                Some(p) => Ok(p.into()),
+                None => Err(e),
+            },
```

Comment:
I was a bit confused reading this but tell me if I understand this correctly: if we pass `"-"` inside `files` then `WalkDir` will give us an error; however, we _can_ turn `"-"` into a valid `PathBuf` by doing `e.path().into()`?

Reply:
Yes, that's correct. We propagate all paths that `WalkDir` wasn't able to deal with to `read_single_file` function. I agree this is a bit clever and confusing but I wasn't able to find a better way.

Reply:
What about the following approach: why don't we separate out any strings that we know to have special semantics and chain them on the end of the iterated files? E.g.:
```
let (normal_files, special_files) = drain_specials(files);
for file in iterate_files(normal_files).chain(special_files)
```
Where `drain_specials` could be something like (there's surely a better way):
```
/// This could use `drain_filter` instead but it is a nightly-only experimental API.
fn drain_specials(mut files: Vec<String>) -> (Vec<String>, Vec<String>) {
    let mut normals = vec![];
    let mut specials = vec![];
    for s in files.drain(..) {
        match s.as_str() {
            "-" => specials.push(s),
            _ => normals.push(s),
        }
    }
    (normals, specials)
}
```
The reason I propose something like this is to avoid the catch-and-then-use of `e.path().into()` which feels hard to debug (and to read too, as you mention).

Reply:
Does it make sense to have multiple stdin inputs? For example with `clif-util run - -` user fills in first input and then second. If not, then we can check if stdin arg exist, filter out all the `-` from `files` and `chain` the `-` at the end as you have proposed.

Reply:
Yeah, to me deduplicating special things like `-` makes sense.